src/NUnitConsole/nunit-console/Utilities/TeamCityServiceMessagePublisher.cs

```diff
@@ -0,0 +1,389 @@
+// a copy of this software and associated documentation files (the
+// "Software"), to deal in the Software without restriction, including
+// without limitation the rights to use, copy, modify, merge, publish,
+// distribute, sublicense, and/or sell copies of the Software, and to
+// permit persons to whom the Software is furnished to do so, subject to
+// the following conditions:
+// 
+// The above copyright notice and this permission notice shall be
+// included in all copies or substantial portions of the Software.
+// 
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
+// LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+// WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+// ***********************************************************************
+
+using System.IO;
+using System.Xml;
+using System.Collections.Generic;
+using System.Globalization;
+
+namespace NUnit.ConsoleRunner.Utilities
+{
+    using System;
+
+    internal class TeamCityServiceMessagePublisher
+    {
+        private readonly TextWriter _outWriter;
+        private readonly Dictionary<string, string> _refs = new Dictionary<string, string>();
+        private int _blockCounter;
+        private string _rootFlowId;
+
+        public TeamCityServiceMessagePublisher(TextWriter outWriter)
+        {
+            if (outWriter == null)
+            {
+                throw new ArgumentNullException("outWriter");
+            }
+
+            _outWriter = outWriter;
+        }
+
+        public void RegisterMessage(XmlNode message)
+        {
+            if (message == null)
+            {
+                throw new ArgumentNullException("message");
+            }
+
+            var messageName = message.Name;
+            if (string.IsNullOrEmpty(messageName))
+            {
+                return;
+            }
+
+            messageName = messageName.ToLowerInvariant();
+            if (messageName == "start-run")
+            {
+                _refs.Clear();
+                return;
+            }           
+
+            var fullName = message.GetAttribute("fullname");
+            if (string.IsNullOrEmpty(fullName))
+            {
+                return;
+            }
+
+            var id = message.GetAttribute("id");
+            var parentId = message.GetAttribute("parentId");
+            string flowId;
+            if (parentId != null)
+            {
+                // NUnit 3 case
+                string rootId;
+                flowId = TryFindRootId(parentId, out rootId) ? rootId : id;
+            }
+            else
+            {
+                // NUnit 2 case
```

Comment:
There is no way to distinguish NUnit 2 from NUnit 3 output, by design. I haven't read far enough to see how you got a parentid into the output for NUnit 3 only, so I'll comment further when I find it. However, since anyone can write a driver extension, I think it's essential that we demand all drivers produce identical messages for the engine.